gui/widgets/grid.cpp

```diff
@@ -0,0 +1,640 @@
+							ThemeEngine::kFontColorAlternate, false);
+		}
+	} else
+		g_gui.theme()->drawSurface(Common::Point(_x, _y), _thumbGfx, true);
+
+	// Draw Platform Icon
+	const Graphics::ManagedSurface *platGfx = _grid->platformToSurface(_activeEntry->platform);
+	if (platGfx) {
+		g_gui.theme()->drawSurface(Common::Point(_x + thumbWidth - 32, _y + thumbHeight - 32), 
+									*platGfx, true);
+	}
+
+	// Draw Flag
+	const Graphics::ManagedSurface *flagGfx = _grid->languageToSurface(_activeEntry->language);
+	if (flagGfx)
+		g_gui.theme()->drawSurface(Common::Point(_x + thumbWidth - flagGfx -> w - 5, _y + 5), 
+									*flagGfx, true);
+
+	// Draw Title
+	if (_grid->_isTitlesVisible) {
+		if (titleLines.size() < 2) {
+			titleLines.push_back(U32String());
+		} else if (titleLines.size() > 2) {
+			titleLines[1].erase(titleLines[1].size() - 3);
+			titleLines[1] += U32String("...");
+		}
+		g_gui.theme()->drawText(Common::Rect(_x, _y + thumbHeight, _x + thumbWidth, _y + thumbHeight + kLineHeight),
+								titleLines[0], GUI::ThemeEngine::kStateEnabled, Graphics::kTextAlignCenter);
+		g_gui.theme()->drawText(Common::Rect(_x, _y + thumbHeight + kLineHeight, _x + thumbWidth, _y + thumbHeight + 2 * kLineHeight),
+								titleLines[1], GUI::ThemeEngine::kStateEnabled, Graphics::kTextAlignCenter);
+	}
+}
+
+void GridItemWidget::handleMouseWheel(int x, int y, int direction) {
+	_grid->handleMouseWheel(x, y, direction);
+	isHighlighted = false;
+}
+
+void GridItemWidget::handleMouseEntered(int button) {
+	if (!isHighlighted) {
+		isHighlighted = true;
+		markAsDirty();
+	}
+}
+
+void GridItemWidget::handleMouseLeft(int button) {
+	if (isHighlighted) {
+		isHighlighted = false;
+		markAsDirty();
+	}
+}
+
+void GridItemWidget::handleMouseMoved(int x, int y, int button) {
+	if (!isHighlighted) {
+		handleMouseEntered(button);
+	}
+}
+
+void GridItemWidget::handleMouseDown(int x, int y, int button, int clickCount) {
+	if (isHighlighted && isVisible()) {
+		_grid->_selectedEntry = _activeEntry->entryID;
+		sendCommand(kItemClicked, 0);
+		// Work in progress
+		// Since user expected to click on "entry" and not the "widget", we
+		// must open the tray where the user expects it to be, which might
+		// not be at the new widget location.
+		int oldX = getAbsX(), oldY = getAbsY();
+		int offsetY = 0;
+		if (_y > (_grid->getHeight() - _h - _grid->_trayHeight)) {
+			offsetY = _y - (_grid->getHeight() - _h - _grid->_trayHeight);
+			sendCommand(kSetPositionCmd, _grid->getScrollPos() + offsetY);
+			_grid->markAsDirty();
+			_grid->draw();
+		}
+		_grid->openTray(oldX,  oldY - offsetY + _h, _activeEntry->entryID);
+		_grid->_tray->runModal();
+	}
+}
+
+#pragma mark -
+
+GridItemTray::GridItemTray(GuiObject *boss, int x, int y, int w, int h, int entryID, GridWidget *grid)
+	: Dialog(x, y, w, h), CommandSender(boss) {
+		
+	_entryID = entryID;
+	_boss = boss;
+	_grid = grid;
+	
+	int buttonWidth = w / 3;
```

Comment:
This, and the lines below has a lot of magic numbers.

Reply:
Again, there is really no reason for "3" specifically. The tray just has a hardcoded layout for now for which I have decided buttonWidth to be w / 3. However, as this later on may be accessible through themes, I have created several layout variables which can be later hooked to a theme in commit 6e5e1d3. And I have added a comment explaining this in code as well.